felt/src/lib.rs

```diff
@@ -1052,6 +1088,204 @@ mod test {
             let y = Felt::parse_bytes(y.as_bytes(), 10).unwrap();
             prop_assert!(x.is_multiple_of(&y));
+            // prop_assert!(x.divides(&y));
+        }
+
+        #[test]
+        fn divides_doesnt_panic(ref x in "(0|[1-9][0-9]*)", ref y in "(0|[1-9][0-9]*)") {
+            let x = Felt::parse_bytes(x.as_bytes(), 10).unwrap();
+            let y = Felt::parse_bytes(y.as_bytes(), 10).unwrap();
+            prop_assert!(x.divides(&y));
+        }
+
+        #[test]
+        fn gcd_doesnt_panic(ref x in "(0|[1-9][0-9]*)", ref y in "(0|[1-9][0-9]*)") {
+            let x = Felt::parse_bytes(x.as_bytes(), 10).unwrap();
+            let y = Felt::parse_bytes(y.as_bytes(), 10).unwrap();
+            let gcd1 = x.gcd(&y);
+            let gcd2 = y.gcd(&x);
+            prop_assert_eq!(gcd1, gcd2);
+        }
+
+        #[test]
+        fn is_even(ref x in "(0|[1-9][0-9]*)") {
+            let x = Felt::parse_bytes(x.as_bytes(), 10).unwrap();
+            prop_assert_eq!(x.is_even(), x.to_biguint().is_even());
+        }
+
+        #[test]
+        fn is_odd(ref x in "(0|[1-9][0-9]*)") {
+            let x = Felt::parse_bytes(x.as_bytes(), 10).unwrap();
+            prop_assert_eq!(x.is_odd(), x.to_biguint().is_odd());
+        }
+
+        /// Tests the additive identity of the implementation of Zero trait for felts
+        ///
+        /// ```{.text}
+        /// x + 0 = x       ∀ x
+        /// 0 + x = x       ∀ x
+        /// ```
+        #[test]
+        fn zero_additive_identity(ref x in "(0|[1-9][0-9]*)") {
+            let x = Felt::parse_bytes(x.as_bytes(), 10).unwrap();
+            let zero = Felt::zero();
+            prop_assert_eq!(&x, &(&x + &zero));
+            prop_assert_eq!(&x, &(&zero + &x));
+        }
+
+        /// Tests the multiplicative identity of the implementation of One trait for felts
+        ///
+        /// ```{.text}
+        /// x * 1 = x       ∀ x
+        /// 1 * x = x       ∀ x
+        /// ```
+        #[test]
+        fn one_multiplicative_identity(ref x in "(0|[1-9][0-9]*)") {
+            let x = Felt::parse_bytes(x.as_bytes(), 10).unwrap();
+            let one = Felt::one();
+            prop_assert_eq!(&x, &(&x * &one));
+            prop_assert_eq!(&x, &(&one * &x));
+        }
+        #[test]
+        fn non_zero_felt_is_always_positive(ref x in FELT_NON_ZERO_PATTERN) {
+            let x = Felt::parse_bytes(x.as_bytes(), 10).unwrap();
+            prop_assert!(x.is_positive())
+        }
+
+        #[test]
+        fn felt_is_never_negative(ref x in FELT_PATTERN) {
+            let x = Felt::parse_bytes(x.as_bytes(), 10).unwrap();
+            prop_assert!(!x.is_negative())
+        }
+
+
+        #[test]
+        fn non_zero_felt_signum_is_always_one(ref x in FELT_NON_ZERO_PATTERN) {
+            let x = Felt::parse_bytes(x.as_bytes(), 10).unwrap();
+            let one = Felt::one();
+            prop_assert_eq!(x.signum(), one)
+        }
```

Comment:
IIRC, the `is_positive`, `signum` and `is_negative` for type `Felt` are meant for the signed representation, which reads a field element as the range `(-CAIRO_PRIME/2, CAIRO_PRIME/2)`, where `x > CAIRO_PRIME/2` represents a negative field element. This would make these assumptions wrong.

Reply:
There was the intent to make a separate type for singed field elements to avoid this confusion, I don't exactly remember if it was applied (I'll check later today and come back with a definitive answer).

Reply:
Effectively it still works as I detailed in the first comment.

Reply:
what should be done about them ?

Reply:
I'd say the practical thing is to split each test in two and use a `prop_assume!` to fix the range. One test checks for negatives, one test checks for positives. The `0` case can be made a normal unit test.

Reply:
But that's not what is implemented in `Signed`, do you mean we have to change the implementation and make felts where `x > CAIRO_PRIME/2` negative ? What about having a separate `SignedFelt` type, isn't it a better design than messing with the implementation of `Felt` ?

Reply:
Yes, the idea was to eventually refactor into a `SignedFelt` type that you could build from `Felt`. But for now it is that way. It should be documented and tested as that until we do the refactor. We're aware as it is right now it's unintuitive, but we prioritized fixing quickly fast over having it be beautiful, with our sight on refactoring it afterwards.
If you'd rather have it that way __now__ so tests can be more obvious, we can either see if someone on our side has time to do the refactor now (I think everyone has a task currently, so this is possible but unlikely), or maybe we can create an issue and you can assign it to yourself if you wish. The task would be to track where we actually care about sign (it's a few hints IIRC) as well as create the base shim.